File: polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 20 11:12:01 2019

@author: pratiksha.garode

Model Name: Polynomial Regression
"""

# Date Preprocessing

# Imprting libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing Datasets
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into training set and testing set
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
"""

# =============================================================================
# # Feature scaling
# from sklearn.preprocessing import StandardScaler
# sc_X = StandardScaler()
# X_train = sc_X.fit_transform(X_train)
# X_test = sc_X.transform(X_test)
# =============================================================================

# Fitting the Linear Regression Model to the Dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting the Polynomial Regression to the Dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 5)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualizing the Results of Linear Regression Model
plt.scatter(X, y, color = 'orange')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff(Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualizing the Results of Polynomial Regression Model
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'orange')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff(Polynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Predicting a new Result with Linear Regression
lin_reg.predict(np.array(6.5).reshape(1, 1))

# Predicting a new Result with Polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform(np.array(6.5).reshape(1, 1)))

File: polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting the Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Fitting the Polynomial Regression to the dataset
dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4
dataset$Level5 = dataset$Level ^ 5
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualizing the Results of Linear Regression Model
#install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level , y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Trith or Bluff(Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

# Visualizing the Results of Polynomial Regression Model
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level , y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff(Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')

# Predicting a new Result with Linear Regression
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Predicting a new Result with Polynomial Regression
y_pred = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                               Level2 = 6.5 ^ 2,
                                               Level3 = 6.5 ^ 3,
                                               Level4 = 6.5 ^ 4,
                                               Level5 = 6.5 ^ 5))











